src/game/engine/sndctrl.cpp

```diff
@@ -0,0 +1,198 @@
+{
+    // TODO, take code from Sole, no code in RA
+}
+
+/**
+ * @brief
+ *
+ * @address 0x00550290
+ */
+void SoundControlsClass::Process()
+{
+    static int _tabs[] = { 110, 144, 180 };
+    Adjust_Variables_For_Resolution();
+    MusicListClass musiclist(0, 62, 162, 516, 146, TPF_6PT_GRAD | TPF_NOSHADOW);
+    TextButtonClass okbtn(BUTTON_SOUND_OK, TXT_OK, TPF_CENTER | TPF_NOSHADOW | TPF_6PT_GRAD, 438, 310, 140);
+    ShapeButtonClass stopbtn(BUTTON_SOUND_STOP, GameFileClass::Retrieve("btn-st.shp"), 62, 310);
+    ShapeButtonClass playbtn(BUTTON_SOUND_PLAY, GameFileClass::Retrieve("btn-pl.shp"), 98, 310);
+    TextButtonClass sufflebtn(BUTTON_SOUND_SUFFLE, TXT_OFF, TPF_CENTER | TPF_NOSHADOW | TPF_6PT_GRAD, 222, 310, 50);
+    TextButtonClass repeatbtn(BUTTON_SOUND_REPEAT, TXT_OFF, TPF_CENTER | TPF_NOSHADOW | TPF_6PT_GRAD, 356, 310, 50);
+    SliderClass scorevolsldr(SLIDER_SCORE_VOL, 322, 110, 216, 10, true);
+    SliderClass soundvolsldr(SLIDER_SOUND_VOL, 322, 134, 216, 10, true);
+
+    // dimensions are the same as the dialog itself, this is for capturing input?
+    GadgetClass gadget(28, 54, 584, 292, MOUSE_LEFT_PRESS);
+    // not sure what's this for and why it uses OK button id...
+    ControlClass ctrl(
+        BUTTON_SOUND_OK, 0, 0, g_seenBuff.Get_Width(), g_seenBuff.Get_Height(), MOUSE_LEFT_PRESS | MOUSE_RIGHT_PRESS);
+
+    // set shuffle state
+    sufflebtn.Set_Toggle_Disabled(true);
+    if (Options.Get_Shuffle()) {
+        sufflebtn.Turn_On();
+    } else {
+        sufflebtn.Turn_Off();
+    }
+    sufflebtn.Set_Text(sufflebtn.Get_Toggle_State() ? TXT_ON : TXT_OFF, false);
+
+    // set repeat state
+    repeatbtn.Set_Toggle_Disabled(true);
+    if (Options.Get_Repeat()) {
+        repeatbtn.Turn_On();
+    } else {
+        repeatbtn.Turn_Off();
+    }
+    repeatbtn.Set_Text(repeatbtn.Get_Toggle_State() ? TXT_ON : TXT_OFF, false);
+
+    // set slider states
+    scorevolsldr.Set_Maximum(255);
+    scorevolsldr.Set_Thumb_Size(16);
+    scorevolsldr.Set_Value(Options.Get_Sound_Volume() * 256); // is it that?
+    soundvolsldr.Set_Maximum(255);
+    soundvolsldr.Set_Thumb_Size(16);
+    soundvolsldr.Set_Value(Options.Get_Score_Volume() * 256); // is it that?
+
+    // set graphic page to draw to
+    Set_Logic_Page(g_seenBuff);
+
+    // link all the gadgets
+    GadgetClass *activegdt = &okbtn;
+    musiclist.Add_Tail(*activegdt);
+    stopbtn.Add_Tail(*activegdt);
+    playbtn.Add_Tail(*activegdt);
+    sufflebtn.Add_Tail(*activegdt);
+    repeatbtn.Add_Tail(*activegdt);
+    scorevolsldr.Add_Tail(*activegdt);
+    soundvolsldr.Add_Tail(*activegdt);
+    gadget.Add_Tail(*activegdt);
+    ctrl.Add_Tail(*activegdt);
+
+    // fill up the list entries
+    for (ThemeType i = THEME_FIRST; i < Theme.Max_Themes(); ++i) {
+        if (Theme.Is_Allowed(i)) {
+            int length = Theme.Track_Length(i);
+            const char *name = Theme.Full_Name(i);
+            char *str = new char[100];
```

Comment:
Why are we allocating memory hear instead of having a buffer on the stack with `char str[100];`. There is no delete here so this leaks memory.